src/hooks/useMaciVote.ts

```diff
@@ -0,0 +1,94 @@
+import { publishBatch, type Signer } from "maci-cli/sdk";
+import { useCallback, useState } from "react";
+
+import { config } from "~/config";
+import { useEthersSigner } from "./useEthersSigner";
+import { useSession } from "next-auth/react";
+import { useMaciSignup } from "./useMaciSignup";
+import { useMaciPoll } from "./useMaciPoll";
+
+export interface IVoteArgs {
+  voteOptionIndex: bigint;
+  newVoteWeight: bigint;
+}
+
+export interface IUseMaciVoteData {
+  isLoading: boolean;
+  error?: string;
+  onVote: (
+    args: IVoteArgs[],
+    onError: () => Promise<void>,
+    onSuccess: () => Promise<void>,
+  ) => Promise<void>;
+}
+
+export const useMaciVote = (nonce = 0): IUseMaciVoteData => {
+  const signer = useEthersSigner();
+  const { data } = useSession();
+  const { stateIndex } = useMaciSignup();
+  const { pollData } = useMaciPoll();
+  const [isLoading, setIsLoading] = useState(false);
+  const [error, setError] = useState<string>();
+
+  const onVote = useCallback(
+    async (
+      votes: IVoteArgs[],
+      onError: () => Promise<void>,
+      onSuccess: () => Promise<void>,
+    ) => {
+      if (!signer || !data || !stateIndex || !pollData) {
+        return;
+      }
+
+      if (!votes.length) {
+        await onError();
+        setError("No votes provided");
+        return;
+      }
+
+      const messages = votes.map(
```

Comment:
So what I'm thinking here is the following:
We can fit around 84 votes in a batch (this value could be part of the config) with a a 30M gas limit (need to confirm if on OP's mainnet that's the limit too), but let's take as base 50 votes in a batch. If the user votes for 150 projects that's 150/50 = 3 batches. So this code should take that into account: 
* calculate how many batches
* create arrays with correct nonces
* publish all batches 

150 votes: nonces from 150 to 1
send in order:
* batch1: 150 - 101 
* batch2: 100 - 51
* batch3: 50 -1 

Also, do we allow to resubmit votes once they already voted? Cause that's something else to take into account. (what clr.fund does is fetch all previously sent messages, decrypt them, and resend the previous + the new ones replacing any of the previous if the new ones are for the same vote option index), and for that we would need to have the ECDH key used for encryption so we could decrypt (and I just saw that in the publishMessageBatch sdk command we are not returning it https://github.com/privacy-scaling-explorations/maci/blob/dev/cli/ts/commands/publish.ts#L226)

Reply:
We don't allow to resubmit votes after and I think we need just to use 50 votes max. For the demo I don't think it's a good UX to submit 3 transactions in a row.

Reply:
Right, but when used in prod we will need to because I believe there are around 300 projects every time for rounds. So it's safe to assume one might want to vote > 50 options, and we can only do it all in one go as the last message will need to have the lowest nonce

Reply:
@ctrlc03 I suggest to use only 50 votes per ballot for one user. I don't think it will be the case when someone really wants to vote for each 300 projects. So, let's add a limit for votes and we don't have to think about UX here because it creates more problems than it solves (like what if first transaction is confirmed and second one is failed).

Reply:
sounds like too much of a limitation to me to only allow voting for 50 projects

Reply:
Well, it's better than submit 6 transactions in a row. I thought the idea of batch method to prevent submitting multiple transactions because it's not convenient for user to confirm them and from dev side to handle all these transactions.

Reply:
Also we should consider that there will be several occasions where this case (multiple TXs) is not even triggered, because a user is within the limit of the batch, so bad ux will only happen in certain occasions

Reply:
In this case I don't understand why we need to support this edge cases for demo app.

Reply:
My bad for wording it that way, I wouldn't consider it a very unlikely case, especially in the context of the OP rpgf. Would you be happy to put this feature in at this time?

Reply:
For the sake of demo, decided to add vote limit and figure out solid solution for these cases https://github.com/privacy-scaling-explorations/maci-rpgf/issues/31.